pkg/controller/component/storaged_updater.go

```diff
@@ -0,0 +1,320 @@
+			return false
+		}
+		if time.Now().After(transLeaderBeginTime.Add(TransLeaderTimeout)) {
+			log.Error(err, "trans leader timeout")
+			return true
+		}
+	}
+	return false
+}
+
+func (s *storagedUpdate) transLeaderIfNecessary(
+	nc *v1alpha1.NebulaCluster,
+	mc nebula.MetaInterface,
+	transAddr string,
+	updatePod *corev1.Pod,
+) error {
+	log := getLog().WithValues("namespace", nc.GetNamespace(), "name", nc.GetName())
+	sc, err := s.getReadyStorageClient(nc)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := sc.Disconnect(); err != nil {
+			log.Error(err, "storage client disconnect failed")
+		}
+	}()
+
+	spaceItems, err := mc.GetSpaceParts()
+	if err != nil {
+		log.Error(err, "get space partition failed")
+		return err
+	}
+	for spaceID, partItems := range spaceItems {
+		for _, partItem := range partItems {
+			if partItem.Leader == nil {
+				continue
+			}
+			if partItem.Leader.Host == transAddr {
+				newLeader := getNewLeader(partItem.Leader, partItem.Peers)
+				if err := s.transLeader(sc, nc, spaceID, partItem.PartID, newLeader, updatePod); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (s *storagedUpdate) transLeader(
+	storageClient nebula.StorageInterface,
+	nc *v1alpha1.NebulaCluster,
+	spaceID ng.GraphSpaceID,
+	partID ng.PartitionID,
+	newLeader *ng.HostAddr,
+	pod *corev1.Pod,
+) error {
+	log := getLog().WithValues("namespace", nc.GetNamespace(), "name", nc.GetName())
+	if err := storageClient.TransLeader(spaceID, partID, newLeader); err != nil {
+		return err
+	}
+	log.Info("transfer leader successfully", "space", spaceID, "partition", partID,
+		"pod", pod.GetName())
+	return nil
+}
+
+func (s *storagedUpdate) getReadyStorageClient(nc *v1alpha1.NebulaCluster) (nebula.StorageInterface, error) {
+	log := getLog().WithValues("namespace", nc.GetNamespace(), "name", nc.GetName())
+	namespace := nc.GetNamespace()
+	storageSvcName := nc.StoragedComponent().GetServiceName()
+
+	storageEndpoint, err := s.clientSet.Endpoint().GetEndpoints(namespace, storageSvcName)
+	if err != nil {
+		log.Error(err, "get storaged endpoints failed")
+		return nil, err
+	}
+
+	eps := getReadyAddress(storageEndpoint)
+	sc, err := nebula.NewStorageClient(eps)
+	if err != nil {
+		transErr, ok := err.(thrift.TransportException)
+		if !ok {
+			return nil, err
+		}
+		if transErr.TypeID() != thrift.NOT_OPEN {
+			return nil, transErr
+		}
+		var retryErr error
+		for retry := 0; retry < len(eps); retry++ {
+			// remove unreachable endpoint
+			eps = append(eps[:0], eps[1:]...)
+			sc, retryErr = nebula.NewStorageClient(eps)
+			if retryErr == nil {
+				break
+			}
+		}
+	}
+	return sc, nil
+}
+
+func getNewLeader(leader *ng.HostAddr, peers []*ng.HostAddr) *ng.HostAddr {
+	others := make([]*ng.HostAddr, 0)
+	for _, peer := range peers {
+		if leader.Host == peer.Host {
+			continue
+		}
+		others = append(others, peer)
+		return others[rand.Intn(len(others))]
```

Comment:
Is the `others` always only one element(`peer`) ?

Reply:
Refactored